### ISSUE TITLE
Performance problems with "Assign Role" or "Remove Assigned Role"

### ISSUE THREAD

**Issue by posixeleni** [IQSS/dataverse#2036]
When we try to remove an Assigned Role it spins and when we refresh the page it does nothing. Discovered while meeting with @mcrosas when looking into AJPS.

**Comment by pdurbin** [comment]
On my laptop on commit ab0b268 (pre-4.1 release) I loaded up production data and played around with granting and revoking a role on a dataverse ("togo") with 32 datasets and 194 files. To me it's unacceptably slow even with Solr is down, taking nearly 10 seconds:

 | Solr Up  | Solr Down
--- | --- | ---
grant  | 0m25.653s | 0m9.699s
revoke  | 0m23.539s | 0m9.968s

Obviously, at ~25 seconds it's crazy slow when Solr is up, which is related to #50.

I'm not sure what our target is for response time for the operation to complete but I guess we'll try to make it as fast as we can. ab0b268 contains the script I'm testing with.

**Comment by pdurbin** [comment]
Based on the profiler (screenshot below), I'm pretty sure the slowdowns are in:

- instantiating datasets (and possibly files and dataverses): DvObjectServiceBean.findDvObject()
- committing changes to Solr over and over: SolrServer.commit()

I'm not sure how to make the instantiation of datasets faster. This issue was noted by @raprasad in #1822 and he suggested storing a version of the dataset as JSON. I could dig into query counts like he did but I trust him. :)

As for repetitive commits to Solr, I could look into batching the operations, similar in spirit to the [batching we've talked about in the past](https://docs.google.com/a/harvard.edu/document/d/12nULSSB4qArA-cYL297o0DibQXuhlJ4J7gkVJVokL-M/edit?usp=sharing).

![screen shot 2015-07-15 at 2 28 22 pm](https://cloud.githubusercontent.com/assets/21006/8706462/d8cf549e-2afd-11e5-8b57-27cdd339f866.png)

**Comment by bencomp** [comment]
@pdurbin is an index created automatically for primary keys?

**Comment by pdurbin** [comment]
In 75e8dba in a new branch I reduced the number of calls to findDvObject but I'd like to review it with @scolapasta before merging it. See the long commit message for details but it end with **much** shorter times than what I reported at https://github.com/IQSS/dataverse/issues/2036#issuecomment-121332191 before I started working on this issue. The numbers are from running scripts/issues/2036/grant-role-then-revoke

            | Solr Up   | Solr Down
    ------- | --------- | -------
    grant   | 0m7.422s  | 0m10.348s
    revoke  | 0m6.758s  | 0m10.341s

What lead me here was the profiler, which now no longer suggests that findDvObject is the culprit. Rather, the next thing to look at would be somehow batching the commits to Solr, but this was not the lower hanging fruit (findDvObject was):

![screen shot 2015-07-16 at 3 58 31 pm](https://cloud.githubusercontent.com/assets/21006/8733982/df69aff2-2bd4-11e5-875c-636fa6b4a863.png)

This screenshot can be compared with the "before" screenshot at https://github.com/IQSS/dataverse/issues/2036#issuecomment-121709925

**Comment by pdurbin** [comment]
@scolapasta as we discussed 75e8dba seemed safe enough to merge into the 4.0.2 branch (which will become 4.1) so I'm passing this to QA to check for any regressions around search permissions.

I'll continue to look at the code locally however and may commit again if I can figure out a good way to batch the Solr commits.

**Comment by kcondon** [comment]
@sbarbosadataverse Can you check this out for performance and pass it back to me to check the other stuff? Thx.

**Comment by sbarbosadataverse** [comment]
@kcondon  I've been dealing with permission performance all week on production so of course I'll check;))

**Comment by pdurbin** [comment]
@sbarbosadataverse thanks for pointing out the real-world performance problems in https://help.hmdc.harvard.edu/Ticket/Display.html?id=218018

**Comment by sbarbosadataverse** [comment]
@pdurbin  should I try this on production or wait for another update to the code/build?

**Comment by pdurbin** [comment]
@sbarbosadataverse I believe @kcondon advised testing on a system with production data.

**Comment by sbarbosadataverse** [comment]
Ok. Thanks

Sonia Barbosa
Manager of Data Curation, IQSS Dataverse Network
Manager of the Murray Research Archive, IQSS
Data Science
Harvard University

Dataverse 4.0 is now available for use!
http://dataverse.harvard.edu

All test dataverses should be created  in 4.0 Demo!
http://dataverse-demo.iq.harvard.edu/

Join our Dataverse Community!
https://groups.google.com/forum/#!forum/dataverse-community
________________________________________

**Comment by pdurbin** [comment]
After noticing how much of a performance hit `getAuthenticatedUser` is at https://github.com/IQSS/dataverse/issues/2036#issuecomment-123447441 (which was also noticed [back in March](https://docs.google.com/a/harvard.edu/document/d/12nULSSB4qArA-cYL297o0DibQXuhlJ4J7gkVJVokL-M/edit?usp=sharing)) I started looking into the roleAssigneeCache @michbarsinai developed in ad87ff8 . By simply commenting out the `resetRoleAssigneeCache()` method per below I'm seeing a fairly dramatic speed increase when granting and revoking roles from the "madagascar" dataverse:

madagascar | reset cache | no reset cache
---------- | ----------- | --------------
grant      | 1m4.935s    | 0m17.516s
revoke     | 1m3.387s    | 0m18.021s

However, I assume `resetRoleAssigneeCache()` was put in for a reason so I have not yet committed the change below. I'll check with @michbarsinai for his thoughts on this.

```
murphy:dataverse pdurbin$ git diff src
diff --git a/src/main/java/edu/harvard/iq/dataverse/search/SearchPermissionsServiceBean.java b/src/main/java/edu/harvard/iq/dataverse/search/SearchPermissionsServiceBean.java
index 9c5289d..90919ae 100644
--- a/src/main/java/edu/harvard/iq/dataverse/search/SearchPermissionsServiceBean.java
+++ b/src/main/java/edu/harvard/iq/dataverse/search/SearchPermissionsServiceBean.java
@@ -85,7 +85,7 @@ public class SearchPermissionsServiceBean {
 
     public List<String> findDvObjectPerms(DvObject dvObject) {
         List<String> permStrings = new ArrayList<>();
-        resetRoleAssigneeCache();
+//        resetRoleAssigneeCache();
         Set<RoleAssignment> roleAssignments = rolesSvc.rolesAssignments(dvObject);
         for (RoleAssignment roleAssignment : roleAssignments) {
             logger.fine("role assignment on dvObject " + dvObject.getId() + ": " + roleAssignment.getAssigneeIdentifier());
@@ -97,7 +97,7 @@ public class SearchPermissionsServiceBean {
                 }
             }
         }
-        resetRoleAssigneeCache();
+//        resetRoleAssigneeCache();
         return permStrings;
     }
```

**Comment by sbarbosadataverse** [comment]
@pdurbin 
I tried removing one user from the PSI in internal production site, it took 64 seconds for the "spinning" blue icon to stop, but the user was not successfully removed

**Comment by pdurbin** [comment]
> However, I assume resetRoleAssigneeCache() was put in for a reason so I have not yet committed the change below. I'll check with @michbarsinai for his thoughts on this.

Upon further testing (with my [SearchIT.java integration test](https://github.com/IQSS/dataverse/blob/master/src/test/java/edu/harvard/iq/dataverse/api/SearchIT.java)) I'm realizing that simply uncommenting the `resetRoleAssigneeCache()` method in ae941f1 (a branch) was a Bad Idea. If I run the test twice without redeploying the app in between it fails because stale data (old user IDs from users who have been delete during the first test run) gets saved into Solr. Back to the drawing board to make getAuthenticatedUser faster. :( /cc @michbarsinai

**Comment by michbarsinai** [comment]
Can we make indexing asynchronous? After looking at the code, seems like there's no immediate usage of the indexed result, and users should be OK with having the permissions updated after 2-3 minutes (eventually consistent etc.). As far as I can tell, in most cases, this delay won't event matter to the users.

Additional option, if you have time - test how this works with a  sparse `DvObject` table, rather than the `@Inheritance(strategy=InheritanceType.JOINED)` we currently have.

**Comment by pdurbin** [comment]
I could look into async but speaking of the database does the authenticateduser table have an index on identifier? /cc @raprasad

**Comment by michbarsinai** [comment]
I think so (AuthenticatedUser.java, line 46:)

 @NotNull
    @Column(nullable = false, unique=true)
    private String userIdentifier;

**Comment by pdurbin** [comment]
Ok, and the presence of "Index Scan" on "useridentifier" from `explain` (pro tip from @raprasad ) seems to confirm that it's indexed. Compare with "Seq Scan" for "superuser" below.

I still wonder why `getAuthenticatedUser` shows up so much in the profiler though: https://github.com/IQSS/dataverse/issues/2036#issuecomment-123733813

It's too bad I can't get `roleAssigneeCache` to help me out more.

```
dvndb=# explain select * from authenticateduser where superuser = 't';
                              QUERY PLAN                              
----------------------------------------------------------------------
 Seq Scan on authenticateduser  (cost=0.00..10.20 rows=10 width=3625)
   Filter: superuser
(2 rows)

dvndb=# explain select * from authenticateduser where useridentifier = 'dataverseAdmin';
                                                   QUERY PLAN                                                    
-----------------------------------------------------------------------------------------------------------------
 Index Scan using authenticateduser_useridentifier_key on authenticateduser  (cost=0.00..8.27 rows=1 width=3625)
   Index Cond: ((useridentifier)::text = 'dataverseAdmin'::text)
(2 rows)

dvndb=# select * from pg_indexes where tablename = 'authenticateduser';
 schemaname |     tablename     |              indexname               | tablespace |                                              
    indexdef                                                  
------------+-------------------+--------------------------------------+------------+----------------------------------------------
--------------------------------------------------------------
 public     | authenticateduser | authenticateduser_useridentifier_key |            | CREATE UNIQUE INDEX authenticateduser_useride
ntifier_key ON authenticateduser USING btree (useridentifier)
 public     | authenticateduser | authenticateduser_email_key          |            | CREATE UNIQUE INDEX authenticateduser_email_k
ey ON authenticateduser USING btree (email)
 public     | authenticateduser | authenticateduser_pkey               |            | CREATE UNIQUE INDEX authenticateduser_pkey ON
 authenticateduser USING btree (id)
(3 rows)

dvndb=# 
```

**Comment by pdurbin** [comment]
e5a11cd is a nice speed up for the case when we only have one version of a dataset and it's both the published and latest version.

**Comment by pdurbin** [comment]
0802173 helps a lot because it batches the indexing of files. See the commit message for various speed comparisons but one that I didn't include is that http://hdl.handle.net/1902.1/NGQCIPIDUK was taking 36 seconds to reindex (the whole dataset, not just permissions) and now it takes 2 seconds.

@scolapasta we should probably decide what's fast enough for 4.2. As we discussed, I'll do a build and put it on a server with production data.

**Comment by pdurbin** [comment]
@sbarbosadataverse I just put v. 4.2 build 24-0178b43 on dvn-vm5. Please let me know if the performance is acceptable.

@scolapasta please note that the fixed I've asked you to review at https://github.com/IQSS/dataverse/issues/2453#issuecomment-130807560 (no merged yet) will impact performance.

**Comment by sbarbosadataverse** [comment]
@pdurbin  so far this is much much faster--under 10seconds from what I can count. I'll continue playing with it before closing.

**Comment by sbarbosadataverse** [comment]
working great @pdurbin 
anything else you want me to test?

**Comment by pdurbin** [comment]
@sbarbosadataverse I'm glad to hear it! Again, I'm concerned this will be slowed down again by a fix for #2453 I haven't yet merged so I'm going to give this to @scolapasta to review https://github.com/IQSS/dataverse/issues/2036#issuecomment-132725536 as part of https://github.com/IQSS/dataverse/issues/2453#issuecomment-130807560 .

**Comment by sbarbosadataverse** [comment]
sounds good.

On Wed, Aug 19, 2015 at 4:36 PM, Philip Durbin <notifications@github.com>
wrote:

> @sbarbosadataverse
> <https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_sbarbosadataverse&d=BQMCaQ&c=WO-RGvefibhHBZq3fL85hQ&r=8R6PzVqt1PEocauQgZMGXsGz29-nb19M7eqlo1d8EVs&m=F6DLKnkdxJX4BL8-Bvgu2GZvOt6arXfdouuu_MuBpkg&s=pPee9-IOQDU4Iz3iWmUk4TjHBRCm5a8C6qzkzKL9TR0&e=>
> I'm glad to hear it! Again, I'm concerned this will be slowed down again by
> a fix for #2453
> <https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_IQSS_dataverse_issues_2453&d=BQMCaQ&c=WO-RGvefibhHBZq3fL85hQ&r=8R6PzVqt1PEocauQgZMGXsGz29-nb19M7eqlo1d8EVs&m=F6DLKnkdxJX4BL8-Bvgu2GZvOt6arXfdouuu_MuBpkg&s=hJTfttAb8442vPA27ON22FdM2bqC6lZCI4hEssahd44&e=>
> I haven't yet merged so I'm going to give this to @scolapasta
> <https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_scolapasta&d=BQMCaQ&c=WO-RGvefibhHBZq3fL85hQ&r=8R6PzVqt1PEocauQgZMGXsGz29-nb19M7eqlo1d8EVs&m=F6DLKnkdxJX4BL8-Bvgu2GZvOt6arXfdouuu_MuBpkg&s=MId-1ccTm1OM3OD6Dj2zPoBYhS08bDbXT77E2sZ_AcQ&e=>
> to review #2036 (comment)
> <https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_IQSS_dataverse_issues_2036-23issuecomment-2D132725536&d=BQMCaQ&c=WO-RGvefibhHBZq3fL85hQ&r=8R6PzVqt1PEocauQgZMGXsGz29-nb19M7eqlo1d8EVs&m=F6DLKnkdxJX4BL8-Bvgu2GZvOt6arXfdouuu_MuBpkg&s=Xr6UapFZo-Tjmvdo6OGfrpYlzutpg4Y2QuW_Oe67S1g&e=>
> as part of #2453 (comment)
> <https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_IQSS_dataverse_issues_2453-23issuecomment-2D130807560&d=BQMCaQ&c=WO-RGvefibhHBZq3fL85hQ&r=8R6PzVqt1PEocauQgZMGXsGz29-nb19M7eqlo1d8EVs&m=F6DLKnkdxJX4BL8-Bvgu2GZvOt6arXfdouuu_MuBpkg&s=lNZGShpVJZl5mjXyYxFTYjeQuKCcShMHwnQPBZDXZPk&e=>
> .
>
> —
> Reply to this email directly or view it on GitHub
> <https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_IQSS_dataverse_issues_2036-23issuecomment-2D132773855&d=BQMCaQ&c=WO-RGvefibhHBZq3fL85hQ&r=8R6PzVqt1PEocauQgZMGXsGz29-nb19M7eqlo1d8EVs&m=F6DLKnkdxJX4BL8-Bvgu2GZvOt6arXfdouuu_MuBpkg&s=jlUy0oIRRShmAUEq-iF1vyV9tK9ylyCyOBEasM0aJAQ&e=>
> .
>



-- 
Sonia Barbosa
Manager of Data Curation, IQSS Dataverse Network
Manager of the Murray Research Archive, IQSS
Data Science
Harvard University

Dataverse 4.0 is now available for use!
http://dataverse.harvard.edu

All test dataverses should be created  in 4.0 Demo!
http://dataverse-demo.iq.harvard.edu/

Join our Dataverse Community!
https://groups.google.com/forum/#!forum/dataverse-community

**Comment by pdurbin** [comment]
To fix #2453 I had to merge in a4a37b5 which I expect to result in us losing some of the performance gains made in this issue per my comment at https://github.com/IQSS/dataverse/issues/2453#issuecomment-140482120

I'm passing this ticket to QA. If the performance gains aren't enough, please feel free to give it back to me.

**Comment by kcondon** [comment]
@pdurbin 
OK, test case is to assign 3 users to file downloader role in the PSI/Madagascar dv.
6-8 seconds to look up each name
48 seconds to save perms.
18 seconds to delete a role.

There are 47 files in this dv.

**Comment by kcondon** [comment]
Passing back to Phil for comment on these measurements.

**Comment by pdurbin** [comment]
Sounds like it's still too slow but I'm out of time for 4.2. @scolapasta I'm passing this to you for a decision on this future milestone you'd like someone to keep working on this.

**Comment by pdurbin** [comment]
@scolapasta asked me to create a custom war file to deploy to dvn-vm5 with indexing of permissions disabled when granting or revoking a role which I have placed at http://dvn-vm1.hmdc.harvard.edu/tmp/issues/2036/5ebeb03/dataverse-4.3.war (when deployed, it should say "v. 4.3 build issue2036-5ebeb03" in the lower right). I'm passing this to QA for deployment on dvn-vm5.

In 5ebeb03 (in a branch) you can see the change in the code where `indexDefinitionPoint(assignment.getDefinitionPoint())` is commented out for both grant (save) and revoke.

As we suspected, this line of code is the expensive operation. Before making any changes (on aba7cd2) I ran `scripts/issues/2036/grant-role-then-revoke madagascar` five times and saw heap memory usage on my laptop rise and fall considerably (roughly half a gig or so using the default garbage collector for Java 7) and each "grant" and "revoke" takes about 12 second:

![screen shot 2015-10-13 at 8 28 21 am](https://cloud.githubusercontent.com/assets/21006/10455046/dbb3b162-7186-11e5-8d22-7b9d4156cc53.png)

With the 5ebeb03 version (the `indexDefinitionPoint` call disabled) calling the script five times yield modest increases in heap memory (perhaps 15 MB per call) and each "grant" and "revoke" only takes only a few milliseconds (0m0.044s, for example):

![screen shot 2015-10-13 at 8 35 26 am](https://cloud.githubusercontent.com/assets/21006/10455051/e0d43a68-7186-11e5-8335-25ecaff2758b.png)

**Comment by pdurbin** [comment]
In a meeting we just decided we'll look into using the same `@TransactionAttribute(REQUIRES_NEW)` technique we put in at 896796d for "index all" as part of #50.

**Comment by pdurbin** [comment]
@kcondon the war file build from 209af07 with `@TransactionAttribute(REQUIRES_NEW)` is now available at http://dvn-vm1.hmdc.harvard.edu/tmp/issues/2036/209af07/dataverse-4.3.war

@scolapasta and I were hoping you'd test it on dvn-vm5 but I don't expect a huge improvement in speed (from 13 seconds down to 9 for "madagascar") nor heap memory usage, based on what I'm seeing on my laptop, however.

This is in a new branch so if we like this change we'll need to merge it.

**Comment by kcondon** [comment]
Took average of 10s, did see characteristic 6GB heap allocation/ free. Passing back for consideration.

**Comment by pdurbin** [comment]
@scolapasta I'm passing this to you for advice on what to try next.

As we discussed in a team meeting recently, the nuclear option is to attempt to get away from indexing Solr "permission documents" at all (getting away from the Solr "JOIN" we figured out how to do with [help from the Solr creator](http://lucene.472066.n3.nabble.com/document-level-security-filter-solution-for-Solr-td4126992.html)). Back when I wrote https://trello.com/c/5z5PpR4r/50-design-solr-document-level-security-filter-solution I presented "Option 3: Pass a (potentially long) list of DvObject ids to Solr to filter on" as an undesirable option, an anti-pattern even, from the Solr perspective, but perhaps it's worth investigating. I can write this up as an issue if desired. It would be a lot of work and the calculation of IDs would be done at runtime (search time).

**Comment by pdurbin** [comment]
From talking to @scolapasta I know he'd like me to keep looking at this issue but I may take at least a quick look at #2648 before I do.

**Comment by pdurbin** [comment]
To reflect the fact that I see this issue as lower priority for 4.2.1 than #2648 I'm removing the "Critical" label.

**Comment by pdurbin** [comment]
As we just decided in a meeting, this issue is like #2472 in that we think the performance improvements would be great but we don't want to hold up 4.2.1 so I'm pulling it out of that milestone. Recent changes have all been in branches so there is nothing to QA for 4.2.1. No changes.

**Comment by pdurbin** [comment]
I demoed to @kcondon a fix I just pushed to a separate branch in 1770995 that makes the indexing of permissions asynchronous when granting or revoking roles.

Per the output below, the time to grant was reduced from 14.5 seconds to 0.2 seconds.

@scolapasta I'm going put this in the 4.2.3 milestone so you can review the code and decide if you want it in. I know in the past we've had to back out of async attempts but this one is smaller and more focused so I hope it doesn't cause a problem.

Please note that this commit does **not** attempt to address the "6-8 seconds to look up each name" issues reported above which is an entirely different part of the code. It sounds like @scolapasta @michbarsinai and @landreev chatted about this the other day. Perhaps it should be treated as its own issue.

Below is the before and after of my fix. The madagascar dataverse contains 46 datasets and a total of 472 files as of my testing. I'm not actually sure why this one is so slow. I speculated to @sekmiller that perhaps it's because it has a greater than average number of role assignments.

## BEFORE

```
murphy:dataverse pdurbin$ scripts/issues/2036/grant-role-then-revoke madagascar
Assigning admin to @pdurbin on madagascar...

real    0m14.567s
user    0m0.004s
sys     0m0.004s
{
  "assignee": "@pdurbin",
  "_roleAlias": "admin"
}
Retrieving ID of role to revoke...

real    0m0.075s
user    0m0.003s
sys     0m0.003s
{
  "id": 139263,
  "assignee": "@pdurbin",
  "roleId": 1,
  "_roleAlias": "admin",
  "definitionPointId": 1251
}
Revoking admin from @pdurbin on madagascar...

real    0m13.690s
user    0m0.003s
sys     0m0.003s
"Role Admin revoked for assignee @pdurbin in PSI/Madagascar"
```

## AFTER

```
murphy:dataverse pdurbin$ scripts/issues/2036/grant-role-then-revoke madagascar
Assigning admin to @pdurbin on madagascar...

real    0m0.203s
user    0m0.004s
sys     0m0.004s
{
  "assignee": "@pdurbin",
  "_roleAlias": "admin"
}
Retrieving ID of role to revoke...

real    0m0.095s
user    0m0.003s
sys     0m0.003s
{
  "id": 139264,
  "assignee": "@pdurbin",
  "roleId": 1,
  "_roleAlias": "admin",
  "definitionPointId": 1251
}
Revoking admin from @pdurbin on madagascar...

real    0m0.077s
user    0m0.003s
sys     0m0.003s
"Role Admin revoked for assignee @pdurbin in PSI/Madagascar"
murphy:dataverse pdurbin$ 
```

**Comment by pdurbin** [comment]
@kcondon @scolapasta as we discussed this morning I just merged the 2036-async-grant branch into 4.2.3. It contains a fairly small commit (1770995) that I described previously at https://github.com/IQSS/dataverse/issues/2036#issuecomment-165567932 . In short, the change is make the **indexing** part of changing role assignments asynchronous.

Passing to QA. Some testing around assigning and removing roles would be appreciated, especially paying attention to how long it takes for known slow cases like the madagascar dataverse. I have not looked into memory usage, as previously mentioned at https://github.com/IQSS/dataverse/issues/2036#issuecomment-148228543 as "characteristic 6GB heap allocation/ free". Please note that while " 6-8 seconds to look up each name" has been mentioned at https://github.com/IQSS/dataverse/issues/2036#issuecomment-142393024 this is an entirely different part of the code and probably deserves its own issue. @scolapasta and @landreev discussed that slowness with @michbarsinai when he was recently in town.

**Comment by kcondon** [comment]
6-8s to look up user
2-3s to save
2s-3s to delete
